### PR TITLE
some improvements to bounded vec API

### PR DESCRIPTION
**Pull request by kianenigma**
From my work in https://github.com/paritytech/substrate/tree/kiz-multi-block-playground and #9511, I've worked a lot with bounded vec and have found some issues with it. This is bringing some of them back to master. 

About the `TryCollect`, the downside as of now is that this cannot work for something like `Filter<_, _>`. Instead, we can make a custom marker trait `NonIncreasingIter`, which will mark all iterator types that cannot increase the size of an iterator (almost all of them, right?). Then we can probably support things like `Filter` as well by relying on this custom marker instead of the `ExactSizeIterator`. 

As for the equality update, I think the benefit is clear, and I don't see why we should consider two bounded vecs not equal if they have the same bound, expressed as different types. 

Lastly, I am also very keen on making a `bounded_vec![]` macro as well as it is a pure PITA to construct them frequently otherwise. To the best of my knowledge that would require a proc-macro, so I have hesitated from doing it so far.

### CODE REVIEW

File: frame/support/src/storage/bounded_vec.rs [hunk 1]
```diff
@@ -344,6 +366,35 @@ where
...
+	///
+	/// The length of self must be known in advance (potentially by being an `ExactSizedIterator`),
+	/// and if greater than `Bound`, then `Err` is returned.
+	fn try_collect<Bound: Get<u32>>(self) -> Result<BoundedVec<T, Bound>, Self::Error>;
+}
+
+impl<I, T> TryCollect<T> for I
+where
+	I: ExactSizeIterator,
+	I: Iterator<Item = T>,
```
Comment (bkchr) [inline frame/support/src/storage/bounded_vec.rs:385]:
```suggestion
	I: ExactSizeIterator + Iterator<Item = T>,
```

**Review by bkchr — APPROVED** [review]

**Review by shawntabrizi — APPROVED** [review]

**Comment by shawntabrizi** [comment]
Agree about bounded tree map, but probably we should be looking to remove weak bounded vec versus giving it more features.

**Comment by kianenigma** [comment]
added to btree map and set, but not weak bounded vec.

will wait for one of you to merge or approve again, since the code changed a bit.

**Review by bkchr — APPROVED** [review]

**Comment by kianenigma** [comment]
bot merge

**Pull request merged**